plugins/renepay/pay.c

```diff
@@ -0,0 +1,1798 @@
+
+	// TODO(eduardo): does it make sense to have this limit on attempts?
+	/* I am classifying the flows in attempt cycles. */
+	renepay_new_attempt(renepay);
+	/* We try only MAX_NUM_ATTEMPTS, then we give up. */
+	if ( renepay_attempt_count(renepay) > MAX_NUM_ATTEMPTS)
+	{
+		return renepay_fail(renepay, PAY_STOPPED_RETRYING,
+				    "Reached maximum number of attempts (%d)",
+				    MAX_NUM_ATTEMPTS);
+	}
+
+	struct amount_msat feebudget, fees_spent, remaining;
+
+	if (time_after(time_now(), p->stop_time))
+		return renepay_fail(renepay, PAY_STOPPED_RETRYING, "Timed out");
+
+	/* Total feebudget  */
+	if (!amount_msat_sub(&feebudget, p->maxspend, p->amount))
+	{
+		plugin_err(pay_plugin->plugin,
+			   "%s (line %d) could not substract maxspend=%s and amount=%s.",
+			   __PRETTY_FUNCTION__,
+			   __LINE__,
+			   type_to_string(tmpctx, struct amount_msat, &p->maxspend),
+			   type_to_string(tmpctx, struct amount_msat, &p->amount));
+	}
+
+	/* Fees spent so far */
+	if (!amount_msat_sub(&fees_spent, p->total_sent, p->total_delivering))
+	{
+		plugin_err(pay_plugin->plugin,
+			   "%s (line %d) could not substract total_sent=%s and total_delivering=%s.",
+			   __PRETTY_FUNCTION__,
+			   __LINE__,
+			   type_to_string(tmpctx, struct amount_msat, &p->total_sent),
+			   type_to_string(tmpctx, struct amount_msat, &p->total_delivering));
+	}
+
+	/* Remaining fee budget. */
+	if (!amount_msat_sub(&feebudget, feebudget, fees_spent))
+	{
+		plugin_err(pay_plugin->plugin,
+			   "%s (line %d) could not substract feebudget=%s and fees_spent=%s.",
+			   __PRETTY_FUNCTION__,
+			   __LINE__,
+			   type_to_string(tmpctx, struct amount_msat, &feebudget),
+			   type_to_string(tmpctx, struct amount_msat, &fees_spent));
+	}
+
+	/* How much are we still trying to send? */
+	if (!amount_msat_sub(&remaining, p->amount, p->total_delivering))
+	{
+		plugin_err(pay_plugin->plugin,
+			   "%s (line %d) could not substract amount=%s and total_delivering=%s.",
+			   __PRETTY_FUNCTION__,
+			   __LINE__,
+			   type_to_string(tmpctx, struct amount_msat, &p->amount),
+			   type_to_string(tmpctx, struct amount_msat, &p->total_delivering));
+	}
+
+	// plugin_log(pay_plugin->plugin,LOG_DBG,fmt_chan_extra_map(tmpctx,pay_plugin->chan_extra_map));
+
+	char const * err_msg;
+
+	/* We let this return an unlikely path, as it's better to try once
+	 * than simply refuse.  Plus, models are not truth! */
+	struct pay_flow **pay_flows = get_payflows(
+						renepay,
+						remaining, feebudget,
+
+						/* would you accept unlikely
+						 * payments? */
+						first_time,
+
+				 		/* is entire payment? */
+						amount_msat_eq(p->total_delivering, AMOUNT_MSAT(0)),
+
+						&err_msg);
+
+	// plugin_log(pay_plugin->plugin,LOG_DBG,"get_payflows produced %s",fmt_payflows(tmpctx,pay_flows));
+
+	/* MCF cannot find a feasible route, we stop. */
+	// TODO(eduardo): alternatively we can fallback to `pay`.
```

Comment:
No, that almost always makes things worse :)

Reply:
Got it!